### PR TITLE
Move institutes from config-developer.yml to default extra facets config and add wildcard support for extra facets

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Simplify config-developer.yml by moving the institutes to the default extra facets configuration.

This change is not backwards compatible if you are using a custom config-developer.yml file to define additional institutes. You will have to move the institutes to a new [extra facets configuration file](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#extra-facets) in that case.

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] ~Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)~
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by bouweandela** [comment]
@zklaus Do you think extra documentation is needed for this pull request?

**Comment by bouweandela** [comment]
The documentation build is failing because the scipy documentation appears to be offline. All other tests are green now.

**Comment by valeriupredoi** [comment]
@bouweandela why are we doing this? The config-developer is a solid place for all that jazz, and it's up there in the directory tree, rather than this info buried in three layers of dir structure. It is defo more organized this way, but names like `extra_facets` and information buried in a few subdirs is not exactly user-friendly, given that this info gets edited by others than just us

**Comment by bouweandela** [comment]
@valeriupredoi The extra facets facility provides a standardized way to automatically add extra key-value pairs to datasets defined in the recipe, therefore it makes sense to use it everywhere we do this. This feature is well documented and easily accessible from config-user.yml, see [here](https://esmvaltool--1259.org.readthedocs.build/projects/ESMValCore/en/1259/quickstart/configure.html#extra-facets) for the documentation. That the defaults are buried 3 levels deep in the directory tree doesn't really matter, because we would not expect users to change them. The defaults can simply be overwritten by user-defined files if needed.

**Comment by bouweandela** [comment]
@valeriupredoi Does that answer address your concerns? I really think that it would be much more clear to users if we have one way of adding extra facets instead of two or three.

@zklaus Any chance of this getting a review from you, so it can go into the upcoming release if @valeriupredoi agrees?

**Comment by valeriupredoi** [comment]
cheers @bouweandela - it do make sense yes, sorry I missed yer answer which clears things for me :+1:

**Review by valeriupredoi — APPROVED** [review]

File: doc/quickstart/configure.rst
Comment (zklaus) [inline doc/quickstart/configure.rst:375]: Perhaps we can find a blend of the old and the new description? I think the opening sentence here should make it immediately clear what this is and why it's useful. I think this is now missing a bit.

File: esmvalcore/_config/_config.py
Comment (zklaus) [inline esmvalcore/_config/_config.py:70]: If I understand correctly, this adds the ability to use Unix-style filename expansion for keys in the extra facets.

First, I am a bit uneasy about applying something that is clearly meant for filenames and path to something else. I think at a minimum we should have a comment here, stating clearly that, yes this is for filenames, but we use it for something else, and that works fine and is awesome.

Second, I think this extension of the syntax should be documented.

Third, this is sufficiently different from moving the institutes that I would consider doing this in a separate PR. At a minimum, it requires a change to the PR title, IMHO.

**Review by zklaus — CHANGES_REQUESTED** [review]
I really like this; it makes the config-developer file much clearer both in contents and conceptually. There are small open questions around documentation and versioning.

File: esmvalcore/_config/_config.py
Comment (bouweandela) [reply esmvalcore/_config/_config.py:70]: > If I understand correctly, this adds the ability to use Unix-style filename expansion for keys in the extra facets.

Yes

> First, I am a bit uneasy about applying something that is clearly meant for filenames and path to something else. I think at a minimum we should have a comment here, stating clearly that, yes this is for filenames, but we use it for something else, and that works fine and is awesome.

I also considered using a regular expression, but I felt that the wildcards used for file selection would be more familiar to most users. For example, we also allow these for selecting ancestors in the recipe and tasks from the command line with the `--diagnostics` flag.

> Second, I think this extension of the syntax should be documented.

Good point, I'll work on improving it

> Third, this is sufficiently different from moving the institutes that I would consider doing this in a separate PR. At a minimum, it requires a change to the PR title, IMHO.

I really need a feature like this here, to be able to specify that the institute should be added to any mip and any variable. I'll update the pull request title and description.

Comment (zklaus) [reply esmvalcore/_config/_config.py:70]: > > First, I am a bit uneasy about applying something that is clearly meant for filenames and path to something else. I think at a minimum we should have a comment here, stating clearly that, yes this is for filenames, but we use it for something else, and that works fine and is awesome.
> I also considered using a regular expression, but I felt that the wildcards used for file selection would be more familiar to most users. For example, we also allow these for selecting ancestors in the recipe and tasks from the command line with the `--diagnostics` flag.

I understand completely. In fact, I was rather surprised that the `fnmatch` docs are so filesystem-centric since it seems such a useful feature elsewhere. That's part of what made me weary. But I'll accept that, would just like it to be a bit more explicit for the benefit of a future maintainer/reader, i.e. rename `fnfilter` to something that doesn't imply `f`ile`n`ame and reword the one-line docstring to mention this. I, for example, needed a few minutes to convince myself that this code was not written with filenames in mind.

> > Second, I think this extension of the syntax should be documented.
> Good point, I'll work on improving it
:+1: 

> > Third, this is sufficiently different from moving the institutes that I would consider doing this in a separate PR. At a minimum, it requires a change to the PR title, IMHO.
> I really need a feature like this here, to be able to specify that the institute should be added to any mip and any variable. I'll update the pull request title and description.

If it was me, I'd probably split it up, get the extended syntax in first, and follow immediately with the institute moving. But it's ok.

**Comment by bouweandela** [comment]
@zklaus Could you have another look at this please? I think I addressed all your comments.

File: doc/quickstart/configure.rst
Comment (zklaus) [inline doc/quickstart/configure.rst:590]: Duplicate `for`

**Review by zklaus — APPROVED** [review]
Just a mini typo. Otherwise, great, thanks!

**Comment by zklaus** [comment]
One more thing: Could you please consider if you want to update the title of this PR and add one or more labels?

**Comment by bouweandela** [comment]
Thanks, I updated the title but I couldn't find a relevant label other than the backwards incompatible one. Did you have a particular label in mind?

**Pull request merged**
